File: Dopamine.MiniPlayerModule/ViewModels/CommonMiniPlayerViewModel.vb
Imports Dopamine.Common.Enums
Imports Dopamine.Common.Enums.PlayerEnums
Imports Dopamine.Core.Prism
Imports Dopamine.Core.Settings
Imports Microsoft.Practices.Prism.Commands
Imports Microsoft.Practices.Prism.Mvvm
Imports Microsoft.Practices.Prism.PubSubEvents

Namespace ViewModels
    Public Class CommonMiniPlayerViewModel
        Inherits BindableBase

#Region "Variables"
        Protected mIsPlaylistVisible As Boolean
        Private mIsCoverPlayerChecked As Boolean
        Private mIsMicroPlayerChecked As Boolean
        Private mIsNanoPlayerChecked As Boolean
        Private mIsMiniPlayerAlwaysOnTop As Boolean
        Private mIsMiniPlayerPositionLocked As Boolean
        Private mEventAggregator As IEventAggregator
#End Region

#Region "Commands"
        Public Property ChangePlayerTypeCommand As DelegateCommand(Of String)
        Public Property ToggleMiniPlayerPositionLockedCommand As DelegateCommand
        Public Property ToggleMiniPlayerAlwaysOnTopCommand As DelegateCommand
#End Region

#Region "Properties"
        Public Property IsPlaylistVisible() As Boolean
            Get
                Return mIsPlaylistVisible
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(mIsPlaylistVisible, value)
            End Set
        End Property

        Public Property IsCoverPlayerChecked() As Boolean
            Get
                Return mIsCoverPlayerChecked
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(mIsCoverPlayerChecked, value)
            End Set
        End Property

        Public Property IsMicroPlayerChecked() As Boolean
            Get
                Return mIsMicroPlayerChecked
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(mIsMicroPlayerChecked, value)
            End Set
        End Property

        Public Property IsNanoPlayerChecked() As Boolean
            Get
                Return mIsNanoPlayerChecked
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(mIsNanoPlayerChecked, value)
            End Set
        End Property

        Public Property IsMiniPlayerAlwaysOnTop() As Boolean
            Get
                Return mIsMiniPlayerAlwaysOnTop
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(mIsMiniPlayerAlwaysOnTop, value)
            End Set
        End Property

        Public Property IsMiniPlayerPositionLocked() As Boolean
            Get
                Return mIsMiniPlayerPositionLocked
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(mIsMiniPlayerPositionLocked, value)
            End Set
        End Property
#End Region

#Region "Construction"
        Public Sub New()

            ' Commands
            Me.ChangePlayerTypeCommand = New DelegateCommand(Of String)(Sub(iMiniPlayerType) Me.SetPlayerContextMenuCheckBoxes(CType(iMiniPlayerType, MiniPlayerType)))

            Me.ToggleMiniPlayerPositionLockedCommand = New DelegateCommand(Sub()
                                                                               IsMiniPlayerPositionLocked = Not IsMiniPlayerPositionLocked
                                                                               XmlSettingsClient.Instance.Set(Of Boolean)("Behaviour", "MiniPlayerPositionLocked", IsMiniPlayerPositionLocked)
                                                                           End Sub)

            Me.ToggleMiniPlayerAlwaysOnTopCommand = New DelegateCommand(Sub()
                                                                            IsMiniPlayerAlwaysOnTop = Not IsMiniPlayerAlwaysOnTop
                                                                            XmlSettingsClient.Instance.Set(Of Boolean)("Behaviour", "MiniPlayerOnTop", IsMiniPlayerAlwaysOnTop)
                                                                        End Sub)

            ' Register Commands: all 3 Mini Players need to listen to these Commands, even if 
            ' their Views are not active. That is why we don't use Subscribe and Unsubscribe.
            ApplicationCommands.ChangePlayerTypeCommand.RegisterCommand(Me.ChangePlayerTypeCommand)
            ApplicationCommands.ToggleMiniPlayerPositionLockedCommand.RegisterCommand(Me.ToggleMiniPlayerPositionLockedCommand)
            ApplicationCommands.ToggleMiniPlayerAlwaysOnTopCommand.RegisterCommand(Me.ToggleMiniPlayerAlwaysOnTopCommand)

            'Initialize
            Me.Initialize()
        End Sub
#End Region

#Region "Private"
        Private Sub SetPlayerContextMenuCheckBoxes(iMiniPlayerType As MiniPlayerType)

            Me.IsCoverPlayerChecked = False
            Me.IsMicroPlayerChecked = False
            Me.IsNanoPlayerChecked = False

            Select Case iMiniPlayerType
                Case MiniPlayerType.CoverPlayer
                    Me.IsCoverPlayerChecked = True
                Case MiniPlayerType.MicroPlayer
                    Me.IsMicroPlayerChecked = True
                Case MiniPlayerType.NanoPlayer
                    Me.IsNanoPlayerChecked = True
                Case Else
                    ' Doesn't happen
            End Select
        End Sub

        Private Sub Initialize()

            ' Set the default IsMiniPlayerPositionLocked value
            Me.IsMiniPlayerPositionLocked = XmlSettingsClient.Instance.Get(Of Boolean)("Behaviour", "MiniPlayerPositionLocked")
            Me.IsMiniPlayerAlwaysOnTop = XmlSettingsClient.Instance.Get(Of Boolean)("Behaviour", "MiniPlayerOnTop")

            ' This sets the initial state of the ContextMenu CheckBoxes
            Me.SetPlayerContextMenuCheckBoxes(CType(XmlSettingsClient.Instance.Get(Of Integer)("General", "MiniPlayerType"), MiniPlayerType))
        End Sub
#End Region
    End Class
End Namespace

File: Dopamine.MiniPlayerModule/ViewModels/CommonMiniPlayerViewModel.cs
using Digimezzo.Utilities.Settings;
using Dopamine.Common.Enums;
using Dopamine.Common.Prism;
using Prism.Commands;
using Prism.Events;
using Prism.Mvvm;

namespace Dopamine.MiniPlayerModule.ViewModels
{
    public class CommonMiniPlayerViewModel : BindableBase
    {
        #region Variables
        protected bool isPlaylistVisible;
        private bool isCoverPlayerChecked;
        private bool isMicroPlayerChecked;
        private bool isNanoPlayerChecked;
        private bool isMiniPlayerAlwaysOnTop;
        private bool isMiniPlayerPositionLocked;
        protected IEventAggregator eventAggregator;
        #endregion

        #region Commands
        public DelegateCommand<string> ChangePlayerTypeCommand { get; set; }
        public DelegateCommand ToggleMiniPlayerPositionLockedCommand { get; set; }
        public DelegateCommand ToggleMiniPlayerAlwaysOnTopCommand { get; set; }
        #endregion

        #region Properties
        public bool IsPlaylistVisible
        {
            get { return this.isPlaylistVisible; }
            set { SetProperty<bool>(ref this.isPlaylistVisible, value); }
        }

        public bool IsCoverPlayerChecked
        {
            get { return this.isCoverPlayerChecked; }
            set { SetProperty<bool>(ref this.isCoverPlayerChecked, value); }
        }

        public bool IsMicroPlayerChecked
        {
            get { return this.isMicroPlayerChecked; }
            set { SetProperty<bool>(ref this.isMicroPlayerChecked, value); }
        }

        public bool IsNanoPlayerChecked
        {
            get { return this.isNanoPlayerChecked; }
            set { SetProperty<bool>(ref this.isNanoPlayerChecked, value); }
        }

        public bool IsMiniPlayerAlwaysOnTop
        {
            get { return this.isMiniPlayerAlwaysOnTop; }
            set { SetProperty<bool>(ref this.isMiniPlayerAlwaysOnTop, value); }
        }

        public bool IsMiniPlayerPositionLocked
        {
            get { return this.isMiniPlayerPositionLocked; }
            set { SetProperty<bool>(ref this.isMiniPlayerPositionLocked, value); }
        }
        #endregion

        #region Construction
        public CommonMiniPlayerViewModel()
        {
            // Commands
            this.ChangePlayerTypeCommand = new DelegateCommand<string>(miniPlayerType => this.SetPlayerContextMenuCheckBoxes((MiniPlayerType)(int.Parse(miniPlayerType))));

            this.ToggleMiniPlayerPositionLockedCommand = new DelegateCommand(() =>
            {
                IsMiniPlayerPositionLocked = !IsMiniPlayerPositionLocked;
                SettingsClient.Set<bool>("Behaviour", "MiniPlayerPositionLocked", IsMiniPlayerPositionLocked);
            });

            this.ToggleMiniPlayerAlwaysOnTopCommand = new DelegateCommand(() =>
            {
                IsMiniPlayerAlwaysOnTop = !IsMiniPlayerAlwaysOnTop;
                SettingsClient.Set<bool>("Behaviour", "MiniPlayerOnTop", IsMiniPlayerAlwaysOnTop);
            });

            // Register Commands: all 3 Mini Players need to listen to these Commands, even if 
            // their Views are not active. That is why we don't use Subscribe and Unsubscribe.
            ApplicationCommands.ChangePlayerTypeCommand.RegisterCommand(this.ChangePlayerTypeCommand);
            ApplicationCommands.ToggleMiniPlayerPositionLockedCommand.RegisterCommand(this.ToggleMiniPlayerPositionLockedCommand);
            ApplicationCommands.ToggleMiniPlayerAlwaysOnTopCommand.RegisterCommand(this.ToggleMiniPlayerAlwaysOnTopCommand);

            //Initialize
            this.Initialize();
        }
        #endregion

        #region Private
        private void SetPlayerContextMenuCheckBoxes(MiniPlayerType miniPlayerType)
        {
            this.IsCoverPlayerChecked = false;
            this.IsMicroPlayerChecked = false;
            this.IsNanoPlayerChecked = false;

            switch (miniPlayerType)
            {
                case MiniPlayerType.CoverPlayer:
                    this.IsCoverPlayerChecked = true;
                    break;
                case MiniPlayerType.MicroPlayer:
                    this.IsMicroPlayerChecked = true;
                    break;
                case MiniPlayerType.NanoPlayer:
                    this.IsNanoPlayerChecked = true;
                    break;
                default:
                    break;
                    // Doesn't happen
            }
        }

        private void Initialize()
        {
            // Set the default IsMiniPlayerPositionLocked value
            this.IsMiniPlayerPositionLocked = SettingsClient.Get<bool>("Behaviour", "MiniPlayerPositionLocked");
            this.IsMiniPlayerAlwaysOnTop = SettingsClient.Get<bool>("Behaviour", "MiniPlayerOnTop");

            // This sets the initial state of the ContextMenu CheckBoxes
            this.SetPlayerContextMenuCheckBoxes((MiniPlayerType)SettingsClient.Get<int>("General", "MiniPlayerType"));
        }
        #endregion
    }
}
